### ISSUE TITLE
Debugging React/Webpack test Harness

### ISSUE THREAD

**Issue by gregorskii** [wallabyjs/public#1062]
### Issue description or question

Hi there,

I am having an issue debugging a test harness I am trying to get working. I have the simple react setup working properly, but when I try to import a test helper file that needs to also be compiled by Webpack I get errors in the Wallaby Console.

My goals in this build are to keep the test files near the component jsx file. And to support loading of the test utils that work with React and follow the same import/export conventions, so the test helpers/test harness (at least) need to be piped through webpack.

### Wallaby.js configuration file

Wallaby:
https://github.com/gregorskii/redux-starter/blob/master/wallaby.js

Webpack:
https://github.com/gregorskii/redux-starter/blob/master/build/webpack.test.conf.js

Harness:
https://github.com/gregorskii/redux-starter/blob/master/test/unit/spec/helpers/index.js

Test
https://github.com/gregorskii/redux-starter/blob/master/src/components/app/index-spec.js

Using an autoloader for Karma I also see the error, so this is likely related to webpack. But hoping you can help!

```
> karma start


webpack: Compiled successfully.
webpack: Compiling...

WARNING in ./~/jsdom/lib/jsdom/living/xmlhttprequest.js
20:23-30 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

WARNING in ./~/ajv/lib/async.js
96:20-33 Critical dependency: the request of a dependency is an expression

WARNING in ./~/ajv/lib/async.js
119:15-28 Critical dependency: the request of a dependency is an expression

WARNING in ./~/ajv/lib/compile/index.js
13:21-34 Critical dependency: the request of a dependency is an expression

ERROR in ./~/jsdom/lib/jsdom.js
Module not found: Error: Can't resolve 'fs' in '/node_modules/jsdom/lib'
 @ ./~/jsdom/lib/jsdom.js 6:11-24
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/jsdom/lib/jsdom/browser/resource-loader.js
Module not found: Error: Can't resolve 'fs' in '/node_modules/jsdom/lib/jsdom/browser'
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js 7:11-24
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/tough-cookie/lib/cookie.js
Module not found: Error: Can't resolve 'net' in '/node_modules/tough-cookie/lib'
 @ ./~/tough-cookie/lib/cookie.js 32:10-24
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/jsdom/lib/jsdom/living/xmlhttprequest.js
Module not found: Error: Can't resolve 'child_process' in '/node_modules/jsdom/lib/jsdom/living'
 @ ./~/jsdom/lib/jsdom/living/xmlhttprequest.js 4:18-42
 @ ./~/jsdom/lib/jsdom/browser/Window.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/jsdom/lib/jsdom/living/xhr-utils.js
Module not found: Error: Can't resolve 'fs' in '/node_modules/jsdom/lib/jsdom/living'
 @ ./~/jsdom/lib/jsdom/living/xhr-utils.js 5:11-24
 @ ./~/jsdom/lib/jsdom/living/xmlhttprequest.js
 @ ./~/jsdom/lib/jsdom/browser/Window.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/cssstyle/lib/CSSStyleDeclaration.js
Module not found: Error: Can't resolve 'fs' in '/node_modules/cssstyle/lib'
 @ ./~/cssstyle/lib/CSSStyleDeclaration.js 7:9-22
 @ ./~/jsdom/lib/jsdom/browser/Window.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/request/lib/har.js
Module not found: Error: Can't resolve 'fs' in '/node_modules/request/lib'
 @ ./~/request/lib/har.js 3:9-22
 @ ./~/request/request.js
 @ ./~/request/index.js
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/forever-agent/index.js
Module not found: Error: Can't resolve 'net' in '/node_modules/forever-agent'
 @ ./~/forever-agent/index.js 6:10-24
 @ ./~/request/request.js
 @ ./~/request/index.js
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/forever-agent/index.js
Module not found: Error: Can't resolve 'tls' in '/node_modules/forever-agent'
 @ ./~/forever-agent/index.js 7:10-24
 @ ./~/request/request.js
 @ ./~/request/index.js
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/tunnel-agent/index.js
Module not found: Error: Can't resolve 'net' in '/node_modules/tunnel-agent'
 @ ./~/tunnel-agent/index.js 3:10-24
 @ ./~/request/lib/tunnel.js
 @ ./~/request/request.js
 @ ./~/request/index.js
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js

ERROR in ./~/tunnel-agent/index.js
Module not found: Error: Can't resolve 'tls' in '/node_modules/tunnel-agent'
 @ ./~/tunnel-agent/index.js 4:10-24
 @ ./~/request/lib/tunnel.js
 @ ./~/request/request.js
 @ ./~/request/index.js
 @ ./~/jsdom/lib/jsdom/browser/resource-loader.js
 @ ./~/jsdom/lib/jsdom.js
 @ ./test/unit/spec/helpers/index.js
 @ ./src/components/app/index-spec.js
 @ ./src -spec\.js(x)?$
 @ ./test/unit/spec/index.js
webpack: Failed to compile.
10 03 2017 17:40:49.184:INFO [karma]: Karma v1.5.0 server started at http://0.0.0.0:9876/
10 03 2017 17:40:49.187:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
10 03 2017 17:40:49.193:INFO [launcher]: Starting browser PhantomJS
10 03 2017 17:40:49.977:INFO [PhantomJS 2.1.1 (Mac OS X 0.0.0)]: Connected on socket DIqMTREdljijtRzXAAAA with id 84298755
PhantomJS 2.1.1 (Mac OS X 0.0.0) ERROR
  ReferenceError: Can't find variable: React
  at undefined:1
PhantomJS 2.1.1 (Mac OS X 0.0.0) ERROR
  ReferenceError: Can't find variable: React
  at undefined:1

PhantomJS 2.1.1 (Mac OS X 0.0.0): Executed 0 of 0 ERROR (0.296 secs / 0 secs)


npm ERR! Darwin 15.6.0
npm ERR! argv ".nvm/versions/node/v7.7.2/bin/node" ".nvm/versions/node/v7.7.2/bin/npm" "run" "test"
npm ERR! node v7.7.2
npm ERR! npm  v4.1.2
npm ERR! code ELIFECYCLE
npm ERR! redux-starter@1.0.0 test: `karma start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the redux-starter@1.0.0 test script 'karma start'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the redux-starter package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     karma start
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs redux-starter
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls redux-starter
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /npm-debug.log
```

It manifests itself in Atom Wallaby similar:

```
SyntaxError: Unexpected token ')'
SyntaxError: Unexpected token ')'
at http://localhost:50687/__modules/287.js?1489196403944:1
Module not found: Error: Can't resolve 'fs' in '/node_modules/jsdom/lib'Error
  at factoryCallback (/node_modules/webpack/lib/Compilation.js:259:39)
....
```

### Code editor or IDE name and version

Atom v1.x

### OS name and version

OSX

**Comment by ArtemGovorov** [comment]
The error you've reported seems to be about that webpack can't bundle `jsdom`. To be honest I'm a bit confused about the setup. You're using webpack and seems like trying to make your tests run in a browser environment, but yet are using jsdom that is designed to simulate browser environment in node.

I have sent you [the pull request](https://github.com/gregorskii/redux-starter/pull/1) to make your wallaby config work and run test in [node environment](https://wallabyjs.com/docs/integration/node.html), similar to how jest works. You don't need webpack to do it, because node natively supports CommonJs (that your babel config compiles you ES6 modules to) after I add the [babel compiler that was missing](https://wallabyjs.com/docs/integration/es-next.html) to your [wallaby config](https://github.com/gregorskii/redux-starter/pull/1/files#diff-4e87b2c6dcfa10beeaf6797eaa6c66ecR26). I have also 
- cleaned up the `files` patterns a bit, added `.css/.scss` files to the patterns (as you seem to use it, so will need to add), 
- added the [node modules hint](https://github.com/gregorskii/redux-starter/pull/1/files#diff-4e87b2c6dcfa10beeaf6797eaa6c66ecR6) (because you seem to set up module roots in your webpack config this way to be able to resolve without relative paths), 
- added the [commonly used stub for the style files in node](https://github.com/gregorskii/redux-starter/pull/1/files#diff-4e87b2c6dcfa10beeaf6797eaa6c66ecR30).

Now this works in node (and many people these days test their React code in node via jest/mocha, as opposed to running tests in browser), and this will not help you to run the tests in karma. To set up karma, or wallaby with webpack like you have been trying, you'll need to remove `jsdom` and perhaps change some other things in your harness. Maybe it'd be easier to setup mocha for node to run your tests similarly to how I configured wallaby (it's what I see many people doing these days).

Hope it helps.

**Comment by gregorskii** [comment]
Ok thank you. I think it was a bit of mixed goals. I was following a redux course on Udemy and trying to make both things work. The test harness uses jsdom to test the components in the browser as his examples have bootstrap and jquery. It was also trying to teach testing as simply as possible, and I was trying to mix in wallaby to learn that as well.

I will review your changes, and appreciate the help. 🙏

**Comment by gregorskii** [comment]
Oh also I had the babel compiler in it originally, I was trying to remove it as I thought it would go through webpack anyways which had babel already. If I set the files to load false, wouldn't babel compile them through webpack?

**Comment by ArtemGovorov** [comment]
> If I set the files to load false, wouldn't babel compile them through webpack?

It would, but [we recommend to remove babel-loader](https://wallabyjs.com/docs/integration/webpack.html#loaders-vs-preprocessorscompilers) from within the wallaby config, as wallaby compiler is more performant.

**Comment by gregorskii** [comment]
Artem,

Did some more work on this and I got it all working. The missing piece for me was using the resolvers in webpack inside babel to keep the same resolvers from my webpack config.

Found this [awesome plugin](https://www.npmjs.com/package/babel-plugin-webpack-alias) that helps to load them into Babel.

These tests are working for now and do not need webpack to run.

I was able to make the CLI tests run by removing Karma for direct running of Mocha.

Thank you for pointing me in the right direction!

For reference here is the commit that fixed the CLI tests for me:

https://github.com/gregorskii/redux-starter/commit/aefd6c444b2dc9027755167c5bd810d3cf90a474
